### ISSUE TITLE
Question: finding SNPs between two groups of samples

### ISSUE THREAD

**Issue by robertwhbaldwin** [nicolaprezza/ebwt2snp-v2#3]
Hello,

I'm trying to think of a way that I can use this tool to detect variants that occur more frequently in one group than another. For example, suppose I have 50 samples sequenced to 20X coverage and there's 25 males and 25 males. I have no reference genome.  The problem is that if I run this tool for each individual sample then I get 50 .snp files. How could I know, for any given SNP, what other samples this SNP is also present in, without doing a lot of pairwise comparisons? Thanks - RObert

**Comment by nicolaprezza** [comment]
Hello Robert,

Assuming that each sample contains only one individual, then by running the tool on individual samples you will only obtain heterozygous sites within that individual (assuming a diploid organism). If you want to find snps between all the samples I suggest to:

1) merge all your samples in one fastq file
2) run the tool on that fastq
3) for each returned snp, align the two sequences on the original dataset (e.g. use bwa or bowtie): this will tell you which samples contain that snp.

**Comment by robertwhbaldwin** [comment]
By "align" the two sequences to the original data set you mean that id:1 and id:2 kmers right? Because I have reads in the snp file that have the same cluster id but are called id:1 and id:2 (example below). And I assume that there can be more than two ids per cluster if your're merging fastq from multiple samples and the position is polymorphic. 

So the advice is to map the kmers for each cluster back to the original reads to determine what samples have what variant. I guess it won't work if you merge the ebwt files together instead ?  

 

>cluster:861_id:1_right:30_cov:5_type:_INDEL_event:A/
ATGGCGGGGGCCACAGCTGTCACTGGTATTAAAAAAAAAAAAAAAAAAAACAAAGAAGGAG
>cluster:861_id:2_right:30_cov:4_type:_INDEL_event:A/
AATGGCGGGGGCCACAGCTGTCACTGGTATTAAAAAAAAAAAAAAAAAAACAAAGAAGGAG

**Comment by nicolaprezza** [comment]
Yes, I meant to map those two kmers against the reads and determine what samples have what variant. There can be more than two ids per cluster. Maybe you'll have to play with the -L and -R options because those two output kmers are probably too long for your reads (e.g. -L 15 -R 15 will generate kmers of length 30 and that should be ok for mapping against reads).

What you are trying to do is actually quite natural but we haven't implemented it in the tool: that would be a generalization of [Mode 3](https://github.com/nicolaprezza/ebwt2InDel#:~:text=Analyze%20two%20datasets%20merged,with%20%220%22%20and%20%221%22.) with a document array allowing arbitrary document IDs (one per sample), rather than just 0/1.

I'm not sure what you mean by "merge the ebwt files together": there's only one ebwt file (first you merge all fastq files into one fastq, then you build one ebwt over this merged fastq).

**Comment by robertwhbaldwin** [comment]
Thanks for your quick reponse Nicola.
If my reads are 100 bp paired end why would 60 bp kmers be too long?
I meant merge the individual sample ebwt files together. At the moment I have 50 of these, one per sample. I guess they're not useful for what I'm trying to do. Couldn't you merge them all together, or merge them into groups (male and female), and then do something with that?

**Comment by nicolaprezza** [comment]
It all depends on the coverage distribution and on the sequencing error rate. The longer your kmer is, the more likely it is that the reads covering the SNP have sequencing errors or they do not fully cover the 60-bp window (if that genome area is not well covered). But with 20x uniform coverage you should be safe I guess ...

Yes, merging the ebwts has the same effect of building one ebwt on the concatenated samples. The tool will still work.

**Comment by robertwhbaldwin** [comment]
Hi Nicola

I merged the individual ebwt files for 43 samples to get one ebwt file (merged.ebwt) and tried to run ebwt2InDel on that file (It works for the individual ebwt files). But I'm getting this error:

/home/robert/tools/ebwt2InDel/build/ebwt2InDel -1 merged.ebwt -L 15 -R 15 -m 4 -o merged_out
This is ebwt2InDel.
Running on one sample (genotyping). Input eBWT file : merged.ebwt
Left-extending eBWT ranges by 15 bases.
Right context length: 15 bases.
Complexity filter: 20
Storing output events to file merged_out
Minimum coverage of output events: 4
Maximum number of variants per genomic position per sample: unlimited.

Phase 1/4: loading and indexing eBWT ... terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)


To merge the ebwt files I just used "cat filea fileb filec ....filex >merged.ebwt". Any ideas why this isn't working? Thanks - Robert

**Comment by nicolaprezza** [comment]
Hi,

The concatenation of two ebwts is not equal to the ebwt of the concatenated
dataset. You should cat the two raw datasets, and then build the ebwt of
the concatenation.

Nicola

Il Gio 7 Ott 2021, 16:37 robertwhbaldwin ***@***.***> ha
scritto:
